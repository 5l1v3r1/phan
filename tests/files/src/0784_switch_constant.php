<?php

namespace NSConstantSwitch;

interface I {
}

class T1 implements I {
    function f1() {
    }
}

class T2 implements I {
    function f2() {
    }
}

class Main {
    /**
     * @param I $arg
     */
    function broken( I $arg ) {
        switch (true) {
            case $arg instanceof T1:
                $arg->f1();
                $arg->f2();  // should warn
                break;
            case $arg instanceof T2:
                $arg->f2(); // PhanUndeclaredMethod Call to undeclared method \I::f2
                $arg->f1();  // should warn
                break;
            case $arg instanceof \Traversable:
            case $arg instanceof \stdClass:
                echo intdiv($arg, 2);  // should infer  \Traversable|\stdClass and warn
                break;
        }
    }
}
