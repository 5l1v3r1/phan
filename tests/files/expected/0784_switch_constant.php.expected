%s:26 PhanUndeclaredMethod Call to undeclared method \NSConstantSwitch\T1::f2 (Did you mean expr->f1())
%s:30 PhanUndeclaredMethod Call to undeclared method \NSConstantSwitch\T2::f1 (Did you mean expr->f2())
%s:34 PhanTypeMismatchArgumentInternalProbablyReal Argument 1 ($numerator) is \Traversable|\stdClass|iterable but \intdiv() takes int
